packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -418,50 +418,76 @@ class ThemeData {
-      cardColor: cardColor,
-      dividerColor: dividerColor,
-      highlightColor: highlightColor,
-      splashColor: splashColor,
-      selectedRowColor: selectedRowColor,
-      unselectedWidgetColor: unselectedWidgetColor,
-      disabledColor: disabledColor,
-      buttonColor: buttonColor,
-      secondaryHeaderColor: secondaryHeaderColor,
-      textSelectionColor: textSelectionColor,
-      textSelectionHandleColor: textSelectionHandleColor,
-      backgroundColor: backgroundColor,
-      dialogBackgroundColor: dialogBackgroundColor,
-      indicatorColor: indicatorColor,
-      hintColor: hintColor,
-      errorColor: errorColor,
-      textTheme: textTheme,
-      primaryTextTheme: primaryTextTheme,
-      accentTextTheme: accentTextTheme,
-      iconTheme: iconTheme,
-      primaryIconTheme: primaryIconTheme,
-      accentIconTheme: accentIconTheme,
-      platform: platform,
+    return new ThemeData.raw(
+      brightness: brightness ?? this.brightness,
+      primaryColor: primaryColor ?? this.primaryColor,
+      primaryColorBrightness: primaryColorBrightness ?? this.primaryColorBrightness,
+      accentColor: accentColor ?? this.accentColor,
+      accentColorBrightness: accentColorBrightness ?? this.accentColorBrightness,
+      canvasColor: canvasColor ?? this.canvasColor,
+      scaffoldBackgroundColor: scaffoldBackgroundColor ?? this.scaffoldBackgroundColor,
+      cardColor: cardColor ?? this.cardColor,
+      dividerColor: dividerColor ?? this.dividerColor,
+      highlightColor: highlightColor ?? this.highlightColor,
+      splashColor: splashColor ?? this.splashColor,
+      selectedRowColor: selectedRowColor ?? this.selectedRowColor,
+      unselectedWidgetColor: unselectedWidgetColor ?? this.unselectedWidgetColor,
+      disabledColor: disabledColor ?? this.disabledColor,
+      buttonColor: buttonColor ?? this.buttonColor,
+      secondaryHeaderColor: secondaryHeaderColor ?? this.secondaryHeaderColor,
+      textSelectionColor: textSelectionColor ?? this.textSelectionColor,
+      textSelectionHandleColor: textSelectionHandleColor ?? this.textSelectionHandleColor,
+      backgroundColor: backgroundColor ?? this.backgroundColor,
+      dialogBackgroundColor: dialogBackgroundColor ?? this.dialogBackgroundColor,
+      indicatorColor: indicatorColor ?? this.indicatorColor,
+      hintColor: hintColor ?? this.hintColor,
+      errorColor: errorColor ?? this.errorColor,
+      textTheme: textTheme ?? this.textTheme,
+      primaryTextTheme: primaryTextTheme ?? this.primaryTextTheme,
+      accentTextTheme: accentTextTheme ?? this.accentTextTheme,
+      iconTheme: iconTheme ?? this.iconTheme,
+      primaryIconTheme: primaryIconTheme ?? this.primaryIconTheme,
+      accentIconTheme: accentIconTheme ?? this.accentIconTheme,
+      platform: platform ?? this.platform,
     );
   }
 
+  static ThemeData _lastBaseTheme;
+  static TextTheme _lastLocalTextGeometry;
+  static ThemeData _memoizedLocalizeResult;
```

Comment:
this is going to be trivial to defeat by just having two themes in the app, which will be pretty common, no?

Did you check if the memoization is actually helpful vs just the aggressive merging?

Reply:
> this is going to be trivial to defeat by just having two themes in the app, which will be pretty common, no?

This optimizes for the case when the number of calls to `Theme.of` is >> the number of `Theme` and `Localizations` widgets. The number of times a theme changes during the build phase of a single frame should be proportional to the number of `Theme`/`Localizations` widgets in the tree. As we build a frame, we do a depth-first traversal (needs citation). `Theme`/`Localizations` applies to a whole sub-tree of widgets. Therefore, while building the sub-tree under a `Theme`/`Localizations` (whichever is the nearest ancestor) we should see a 100% cache hit rate. Only when the build phase moves on to the part of the tree with a different `Theme`/`Localizations`, do we need to re-merge the theme.

> Did you check if the memoization is actually helpful vs just the aggressive merging?

I checked the following options:

- No wrapper, no memoization. Result: even worse that the current regression.
- Keep the wrapper, add memoization: Result: almost no improvement (within the noise).
- No wrapper, add memoization, use `==` instead of `identical`. Result: only gets us halfway back to the original numbers.
- No wrapper, add memoization (this PR): Result: back to the original numbers.

My experiments were not at all scientific. I just ran the benchmark 3 times and eyeballed the numbers.

Reply:
We guarantee that we won't do a node that's a descendant before a node that's an ancestor, but that's as far as we go in terms of guaranteeing an order. We don't generally do depth first except for when the next dirty node happens to be a direct child and there's no layout builders.

I was assuming the optimisation was intended to last across multiple frames, but I guess even if it doesn't it's still a good optimisation in this case. My concern is that this is basically a hidden performance cliff, which we've so far worked really really hard to avoid.

I don't have a better suggestion though...

Reply:
How about an MRU cache of the last N pairs of objects or some such? At least the cliff would be less serious...

Reply:
Added a cache.